File: wordBreak.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool wordBreak(string s, unordered_set<string> &dict) {
        int len = s.size();
        int mark[len];
        int pos = 0;

        for (int i = 0; i < len; ++i) {
            mark[i] = -1;
            for (int j = pos; j >= 0; --j) {
                if (dict.end() != dict.find(s.substr(mark[j]+1, i-mark[j]))) {
                    mark[pos++] = i;
                    break;
                }
            }
        }
        if (pos) return mark[pos-1] == len - 1;
        return false;
    }
};

int main(void) {
    string str = "leetcode";
    unordered_set<string> dict;
    dict.insert("leet");
    dict.insert("code");

    Solution s;
    cout << s.wordBreak(str, dict) << endl;
    return 0;
}

File: go/wordBreak.go
package main

import "fmt"

// https://leetcode-cn.com/problems/word-break/

func wordBreak(s string, wordDict []string) bool {
	N, M := len(s), len(wordDict)
	dict := make(map[string]bool)
	for i := 0; i < M; i++ {
		dict[wordDict[i]] = true
	}

	mark := make([]int, N)
	var search func(start int) bool
	search = func(start int) bool {
		if start >= N {
			return true
		}
		if mark[start] == 1 {
			return true
		}
		if mark[start] == -1 {
			return false
		}
		for i := start + 1; i <= N; i++ {
			if dict[s[start:i]] && search(i) {
				mark[start] = 1
				return true
			}
		}
		mark[start] = -1
		return false
	}
	return search(0)
}

func main() {
	cases := []struct {
		s    string
		dict []string
	}{
		{"leetcode", []string{"leet", "code"}},
		{"applepenapple", []string{"apple", "pen"}},
		{"catsandog", []string{"cats", "dog", "sand", "and", "cat"}},
	}

	realCase := cases[0:]
	for i, c := range realCase {
		fmt.Println("## case", i)
		// solve
		fmt.Println(wordBreak(c.s, c.dict))
	}
}
